File: vector_search.py
"""
Ejercicio 4: Buscar un valor en múltiples vectores
Implementación paralela usando concurrent.futures para buscar un valor en vectores independientes
"""

import time
import random
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor
from typing import List, Tuple, Optional
import numpy as np


def search_value_in_vector(vector: Tuple[List[int], int]) -> Tuple[int, bool, int, float]:
    """
    Busca un valor en un vector
    
    Args:
        vector: Tupla con (lista_de_números, valor_a_buscar)
    
    Returns:
        Tupla con (índice_del_vector, encontrado, posición_encontrada, tiempo_procesamiento)
    """
    numbers, target_value = vector
    start_time = time.time()
    
    # Búsqueda lineal
    found = False
    position = -1
    
    for i, num in enumerate(numbers):
        if num == target_value:
            found = True
            position = i
            break
    
    processing_time = time.time() - start_time
    return len(numbers), found, position, processing_time


def search_value_sequential(vectors: List[List[int]], target_value: int) -> List[Tuple[int, bool, int, float]]:
    """
    Versión secuencial de búsqueda en vectores
    
    Args:
        vectors: Lista de vectores donde buscar
        target_value: Valor a buscar
    
    Returns:
        Lista de resultados
    """
    results = []
    
    for i, vector in enumerate(vectors):
        result = search_value_in_vector((vector, target_value))
        results.append((i, result[1], result[2], result[3]))
    
    return results


def search_value_parallel(vectors: List[List[int]], target_value: int, max_workers: int = None) -> List[Tuple[int, bool, int, float]]:
    """
    Versión paralela de búsqueda en vectores usando ProcessPoolExecutor
    
    Args:
        vectors: Lista de vectores donde buscar
        target_value: Valor a buscar
        max_workers: Número máximo de workers
    
    Returns:
        Lista de resultados
    """
    # Preparar datos para procesamiento paralelo
    vector_data = [(vector, target_value) for vector in vectors]
    
    with ProcessPoolExecutor(max_workers=max_workers) as executor:
        results = list(executor.map(search_value_in_vector, vector_data))
    
    return results


def create_sample_vectors(num_vectors: int = 8, vector_size: int = 100000) -> List[List[int]]:
    """
    Crea vectores de muestra para testing
    
    Args:
        num_vectors: Número de vectores a crear
        vector_size: Tamaño de cada vector
    
    Returns:
        Lista de vectores con números aleatorios
    """
    vectors = []
    
    for i in range(num_vectors):
        # Crear vector con números aleatorios
        vector = [random.randint(1, 1000) for _ in range(vector_size)]
        
        # Asegurar que algunos vectores contengan valores específicos para testing
        if i % 3 == 0:  # Cada tercer vector tendrá el valor 42
            vector[random.randint(0, vector_size-1)] = 42
        elif i % 3 == 1:  # Cada segundo vector tendrá el valor 100
            vector[random.randint(0, vector_size-1)] = 100
        
        vectors.append(vector)
    
    return vectors


def benchmark_vector_search():
    """
    Ejecuta benchmark comparando versión secuencial vs paralela
    """
    print("=== EJERCICIO 4: BÚSQUEDA EN MÚLTIPLES VECTORES ===\n")
    
    # Crear vectores de muestra
    print("Creando vectores de muestra...")
    vectors = create_sample_vectors(10, 75000)
    print(f"Se crearon {len(vectors)} vectores de {len(vectors[0])} elementos cada uno\n")
    
    # Valores a buscar
    search_values = [42, 100, 999, 1]
    
    for target_value in search_values:
        print(f"Buscando valor: {target_value}")
        print("-" * 40)
        
        # Benchmark versión secuencial
        print("Ejecutando versión secuencial...")
        start_time = time.time()
        sequential_results = search_value_sequential(vectors, target_value)
        sequential_time = time.time() - start_time
        
        print(f"Tiempo secuencial: {sequential_time:.4f} segundos")
        print(f"Tiempo promedio por vector: {sequential_time/len(vectors):.4f} segundos")
        
        # Contar resultados encontrados
        found_count_sequential = sum(1 for result in sequential_results if result[1])
        print(f"Valores encontrados (secuencial): {found_count_sequential}/{len(vectors)}")
        
        # Benchmark versión paralela
        print("\nEjecutando versión paralela...")
        start_time = time.time()
        parallel_results = search_value_parallel(vectors, target_value)
        parallel_time = time.time() - start_time
        
        print(f"Tiempo paralelo: {parallel_time:.4f} segundos")
        print(f"Tiempo promedio por vector: {parallel_time/len(vectors):.4f} segundos")
        
        # Contar resultados encontrados
        found_count_parallel = sum(1 for result in parallel_results if result[1])
        print(f"Valores encontrados (paralelo): {found_count_parallel}/{len(vectors)}")
        
        # Verificar que los resultados sean iguales
        print(f"\nVerificación de resultados:")
        print(f"Diferencia en cantidad encontrada: {abs(found_count_sequential - found_count_parallel)}")
        
        # Calcular speedup
        speedup = sequential_time / parallel_time if parallel_time > 0 else 0
        print(f"Speedup: {speedup:.2f}x")
        
        # Mostrar resultados detallados
        print("\nResultados detallados:")
        for i, (vector_idx, found, position, proc_time) in enumerate(parallel_results):
            status = f"encontrado en posición {position}" if found else "no encontrado"
            print(f"  Vector {i+1}: {status}, tiempo = {proc_time:.4f}s")
        
        print(f"\nVectores procesados exitosamente: {len(parallel_results)}")
        print("\n" + "="*60 + "\n")


if __name__ == "__main__":
    benchmark_vector_search()

File: vector_search.cpp
/*
 * Ejercicio 4: Buscar un valor en múltiples vectores
 * Implementación paralela usando OpenMP para buscar un valor en vectores independientes
 */

#include <iostream>
#include <vector>
#include <chrono>
#include <random>
#include <omp.h>
#include <iomanip>

using namespace std;

// Estructura para almacenar información de búsqueda en un vector
struct SearchData {
    int vector_id;
    vector<int> data;
    int target_value;
    bool found;
    int position;
    double processing_time;
};

// Función para buscar un valor en un vector
void searchValueInVector(SearchData& search_data) {
    auto start_time = chrono::high_resolution_clock::now();
    
    // Búsqueda lineal
    search_data.found = false;
    search_data.position = -1;
    
    for (size_t i = 0; i < search_data.data.size(); ++i) {
        if (search_data.data[i] == search_data.target_value) {
            search_data.found = true;
            search_data.position = static_cast<int>(i);
            break;
        }
    }
    
    auto end_time = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::microseconds>(end_time - start_time);
    search_data.processing_time = duration.count() / 1000000.0; // Convertir a segundos
}

// Versión secuencial de búsqueda en vectores
vector<SearchData> searchValueSequential(vector<SearchData>& search_data) {
    for (auto& data : search_data) {
        searchValueInVector(data);
    }
    return search_data;
}

// Versión paralela de búsqueda en vectores usando OpenMP
vector<SearchData> searchValueParallel(vector<SearchData>& search_data) {
    #pragma omp parallel for
    for (int i = 0; i < static_cast<int>(search_data.size()); ++i) {
        searchValueInVector(search_data[i]);
    }
    return search_data;
}

// Función para crear vectores de muestra
vector<SearchData> createSampleVectors(int num_vectors = 10, int vector_size = 75000) {
    vector<SearchData> search_data;
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> dis(1, 1000);
    
    for (int i = 0; i < num_vectors; ++i) {
        SearchData data;
        data.vector_id = i + 1;
        
        // Crear vector con números aleatorios
        data.data.reserve(vector_size);
        for (int j = 0; j < vector_size; ++j) {
            data.data.push_back(dis(gen));
        }
        
        // Asegurar que algunos vectores contengan valores específicos para testing
        if (i % 3 == 0) {  // Cada tercer vector tendrá el valor 42
            data.data[dis(gen) % vector_size] = 42;
        } else if (i % 3 == 1) {  // Cada segundo vector tendrá el valor 100
            data.data[dis(gen) % vector_size] = 100;
        }
        
        search_data.push_back(data);
    }
    
    return search_data;
}

// Función principal de benchmark
void benchmarkVectorSearch() {
    cout << "=== EJERCICIO 4: BÚSQUEDA EN MÚLTIPLES VECTORES (C++ con OpenMP) ===" << endl << endl;
    
    // Crear vectores de muestra
    cout << "Creando vectores de muestra..." << endl;
    vector<SearchData> search_data = createSampleVectors(10, 75000);
    cout << "Se crearon " << search_data.size() << " vectores de " << search_data[0].data.size() << " elementos cada uno" << endl << endl;
    
    // Valores a buscar
    vector<int> search_values = {42, 100, 999, 1};
    
    for (int target_value : search_values) {
        cout << "Buscando valor: " << target_value << endl;
        cout << string(40, '-') << endl;
        
        // Configurar datos de búsqueda
        vector<SearchData> current_search_data = search_data;
        for (auto& data : current_search_data) {
            data.target_value = target_value;
            data.found = false;
            data.position = -1;
        }
        
        // Benchmark versión secuencial
        cout << "Ejecutando versión secuencial..." << endl;
        vector<SearchData> sequential_data = current_search_data;
        auto start_time = chrono::high_resolution_clock::now();
        vector<SearchData> sequential_results = searchValueSequential(sequential_data);
        auto end_time = chrono::high_resolution_clock::now();
        auto sequential_duration = chrono::duration_cast<chrono::microseconds>(end_time - start_time);
        double sequential_time = sequential_duration.count() / 1000000.0;
        
        cout << "Tiempo secuencial: " << fixed << setprecision(4) << sequential_time << " segundos" << endl;
        cout << "Tiempo promedio por vector: " << sequential_time / search_data.size() << " segundos" << endl;
        
        // Contar resultados encontrados
        int found_count_sequential = 0;
        for (const auto& result : sequential_results) {
            if (result.found) found_count_sequential++;
        }
        cout << "Valores encontrados (secuencial): " << found_count_sequential << "/" << search_data.size() << endl;
        
        // Benchmark versión paralela
        cout << endl << "Ejecutando versión paralela..." << endl;
        vector<SearchData> parallel_data = current_search_data;
        start_time = chrono::high_resolution_clock::now();
        vector<SearchData> parallel_results = searchValueParallel(parallel_data);
        end_time = chrono::high_resolution_clock::now();
        auto parallel_duration = chrono::duration_cast<chrono::microseconds>(end_time - start_time);
        double parallel_time = parallel_duration.count() / 1000000.0;
        
        cout << "Tiempo paralelo: " << parallel_time << " segundos" << endl;
        cout << "Tiempo promedio por vector: " << parallel_time / search_data.size() << " segundos" << endl;
        
        // Contar resultados encontrados
        int found_count_parallel = 0;
        for (const auto& result : parallel_results) {
            if (result.found) found_count_parallel++;
        }
        cout << "Valores encontrados (paralelo): " << found_count_parallel << "/" << search_data.size() << endl;
        
        // Verificar que los resultados sean iguales
        cout << endl << "Verificación de resultados:" << endl;
        cout << "Diferencia en cantidad encontrada: " << abs(found_count_sequential - found_count_parallel) << endl;
        
        // Calcular speedup
        double speedup = (parallel_time > 0) ? sequential_time / parallel_time : 0;
        cout << "Speedup: " << fixed << setprecision(2) << speedup << "x" << endl;
        
        int num_threads = omp_get_max_threads();
        double efficiency = (speedup / num_threads) * 100;
        cout << "Eficiencia: " << efficiency << "%" << endl;
        cout << "Número de threads utilizados: " << num_threads << endl;
        
        // Mostrar resultados detallados
        cout << endl << "Resultados detallados:" << endl;
        for (const auto& result : parallel_results) {
            string status = result.found ? 
                "encontrado en posición " + to_string(result.position) : 
                "no encontrado";
            cout << "  Vector " << result.vector_id << ": " << status 
                 << ", tiempo = " << result.processing_time << "s" << endl;
        }
        
        cout << endl << "Vectores procesados exitosamente: " << parallel_results.size() << endl;
        cout << endl << string(60, '=') << endl << endl;
    }
}

int main() {
    // Configurar OpenMP
    omp_set_num_threads(omp_get_max_threads());
    
    cout << "OpenMP disponible: " << omp_get_max_threads() << " threads" << endl;
    cout << "Compilador OpenMP: " << _OPENMP << endl << endl;
    
    benchmarkVectorSearch();
    
    return 0;
}
